### PR TITLE
Update modin dependency to 0.30.1

### PR DESCRIPTION
**Pull request by sfc-gh-joshi**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1552497

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

This PR updates modin to 0.30.1, which unpins the pandas patch version and fixes bugs related to the extensions/docstring modules.

Dependency-related changes:
- Moves the pinned modin version from 0.28.1 to 0.30.1--it is available in Snowflake anaconda but currently blocked on prod/preprod, and must be manually enabled for testing accounts in those environments. We will unblock modin 0.30.1 when the release containing this change is completed.
- Adds a new job to the daily test (`test-pandas-patch-versions`), which runs the Snowpark pandas daily test suite on AWS with Python 3.9 against all supported pandas versions currently available in anaconda (2.1 and 2.2; we should add 2.3 once it's available in Snowflake anaconda)
- Some native pandas error messages checked in tests have changed; relevant tests now check the error message based on pandas version.
- **IMPORTANT** The most recent available versions of `snowflake-snowpark-python` available in Anaconda pin the modin dependency to `0.28.1`, so attempting to specify `modin==0.30.1` when creating a stored procedure makes the solver pick an older version of `snowflake-snowpark-python` from before Snowpark pandas was added (1.16.0). ~~As a workaround, in `test_modin_stored_procedures.py` we need to upload `snowflake-snowpark-python` as a staged package. Calling `apply` and `applymap` (which both create UDFs) within a stored procedure created in this way fails; I have XFAILed those tests for now and need help debugging them (either in this PR or as a separate one).~~ Within `test_modin_stored_procedures.py` we pin `pandas==2.2.1` and `modin==0.28.1` to test the version available in Anaconda; other UDF/stored proc-creating tests are XFAILed when the pandas version is 2.2.3 or newer.

Frontend changes:
- modin 0.30.1 introduces the `pd.DataFrame.modin` accessor object for non-pandas methods, such as `pd.DataFrame.modin.to_pandas` and `pd.DataFrame.modin.to_ray`. We will automatically raise NotImplementedError for all methods on this accessor object except `to_pandas`.
- `DataFrame`/`Series.drop_duplicates` was changed to use a different query compiler method to better support distributed backends; for now I've added back an override with the old frontend implementation.
- `Series.unique` was changed to have an additional `to_pandas()` call, which does not affect the number of queries since the old implementation called `to_numpy()`, but does change the return type. I've added back the old frontend implementation as an override.

### CODE REVIEW

**Comment by sfc-gh-azhan** [comment]
When modin 0.30.1 will be available in Snowflake anaconda?

**Comment by sfc-gh-dpetersohn** [comment]
@sfc-gh-azhan should be very soon. Do you know why all the tests would pass even though it is in Anaconda yet?

**Comment by sfc-gh-azhan** [comment]
> @sfc-gh-azhan should be very soon. Do you know why all the tests would pass even though it is in Anaconda yet?

Those tests here are using pypi. Only the Sproc tests or Snowflake streamlit or notebook tests relies on Anaconda.

**Comment by sfc-gh-azhan** [comment]
Once I done with https://snowflakecomputing.atlassian.net/browse/SNOW-1563225, I could run your PR for those sproc tests and I believe they will fail before modin is upgraded in Anaconda.

File: .github/workflows/daily_modin_precommit.yml
Comment (sfc-gh-yzou) [inline .github/workflows/daily_modin_precommit.yml:186]: why do we need to wait for the snwoflake anaconda channel, that impacts stored sproc, notebook, ut shouldn't impact our github daily test

Comment (sfc-gh-yzou) [reply .github/workflows/daily_modin_precommit.yml:186]: but is 2.2.3 the one installed in our merge gate test?

File: .github/workflows/precommit.yml
Comment (sfc-gh-yzou) [inline .github/workflows/precommit.yml:387]: why do we need to pin the pandas version here, is it because we haven't tested with 2.2.3 yet?

File: CHANGELOG.md
Comment (sfc-gh-yzou) [inline CHANGELOG.md:14]: you mean updated pandas support from 2.2.1 to 2.2.*, right?

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/__init__.py:26]: as we discussed before, shall we first check modin, then check pandas?

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/__init__.py:144]: what it that modinAPI? seems something new here

Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/__init__.py:144]: is that a new set of API provided by modin?

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/__init__.py:164]: what is modin accessor object?

File: src/snowflake/snowpark/modin/plugin/extensions/base_overrides.py
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/extensions/base_overrides.py:1556]: where does this part of code coming from? is tht what we use in snowpark pandas before? it is wired that we use result.index for ignore_index, isn't that just a reset index?

File: src/snowflake/snowpark/modin/plugin/extensions/series_overrides.py
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/extensions/series_overrides.py:1598]: is this counted as a bug in modin?

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-yzou) [inline tests/integ/modin/test_modin_stored_procedures.py:21]: do we need to import that from snowflake.snowpark.modin.plugin? can we just get the current pandas version here?

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-yzou) [inline tests/integ/modin/test_modin_stored_procedures.py:37]: this is to upload the current snowpark code to stored sproc, right?

File: .github/workflows/daily_modin_precommit.yml
Comment (sfc-gh-joshi) [reply .github/workflows/daily_modin_precommit.yml:188]: What's the merge gate test you're referring to? When I tried adding 2.2.3, all tests that used a UDF or sproc failed because it's not in anaconda (you can see an example in the daily job run where I forgot to pin the pandas version: https://github.com/snowflakedb/snowpark-python/actions/runs/11285715088/job/31389018772).

File: .github/workflows/precommit.yml
Comment (sfc-gh-joshi) [reply .github/workflows/precommit.yml:387]: as above--not pinning makes UDF/sprocs fail

File: CHANGELOG.md
Comment (sfc-gh-joshi) [reply CHANGELOG.md]: yes, I'll update this to clarify

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-joshi) [reply src/snowflake/snowpark/modin/plugin/__init__.py:144]: yes, this `ModinAPI` is responsible for this part I mentioned in the PR description:
> modin 0.30.1 introduces the pd.DataFrame.modin accessor object for non-pandas methods, such as pd.DataFrame.modin.to_pandas and pd.DataFrame.modin.to_ray. We will automatically raise NotImplementedError for all methods on this accessor object except to_pandas.

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-joshi) [reply src/snowflake/snowpark/modin/plugin/__init__.py:164]: it's methods defined on the `ModinAPI` object mentioned in the [above comment](https://github.com/snowflakedb/snowpark-python/pull/1965/files#r1797211823)

File: src/snowflake/snowpark/modin/plugin/extensions/base_overrides.py
Comment (sfc-gh-joshi) [reply src/snowflake/snowpark/modin/plugin/extensions/base_overrides.py:1556]: yes, I copied this code exactly as it was from 0.28.1: https://github.com/modin-project/modin/blob/eac21a8463d7152b3ad48a8b80d6da5ac645be6a/modin/pandas/base.py#L1503-L1528

I'm not sure why this index set is here intead of a reset_index either, but I just kept it as-is to avoid any potentially breaking changes

File: src/snowflake/snowpark/modin/plugin/extensions/series_overrides.py
Comment (sfc-gh-joshi) [reply src/snowflake/snowpark/modin/plugin/extensions/series_overrides.py:1598]: It's a bug for both Snowpark pandas and modin, just in slightly different ways. The expected native pandas behavior is to always return an `Index` class; Snowpark pandas currently returns `np.array` (tracked by SNOW-1019312), while modin after 0.30.1 returns a pandas `DatetimeArray` instead.

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-joshi) [reply tests/integ/modin/test_modin_stored_procedures.py:37]: yep

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-joshi) [reply tests/integ/modin/test_modin_stored_procedures.py:21]: that would work as well, do you have a preference for either way?

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-joshi) [reply src/snowflake/snowpark/modin/plugin/__init__.py:26]: I tried this approach (importing modin, catching the warning it raises, then re-raising an error), but it seems like modin will sometimes raise both an error AND a warning. I tried installing `pandas==1.5.3` and running `import modin.pandas`, and this is the error it gave:
```
>>> import modin.pandas
UserWarning: The pandas version installed (1.5.3) does not match the supported pandas version in Modin (2.2.X). This may cause undesired side effects!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/joshi/miniconda3/envs/snowpandas-modin-0.30.1/lib/python3.9/site-packages/modin/pandas/__init__.py", line 178, in <module>
    from .dataframe import DataFrame
  File "/Users/joshi/miniconda3/envs/snowpandas-modin-0.30.1/lib/python3.9/site-packages/modin/pandas/dataframe.py", line 64, in <module>
    from modin.pandas.io import from_non_pandas, from_pandas, to_pandas
  File "/Users/joshi/miniconda3/envs/snowpandas-modin-0.30.1/lib/python3.9/site-packages/modin/pandas/io.py", line 50, in <module>
    from pandas._typing import (
ImportError: cannot import name 'DtypeBackend' from 'pandas._typing' (/Users/joshi/miniconda3/envs/snowpandas-modin-0.30.1/lib/python3.9/site-packages/pandas/_typing.py)
```

This can be worked around (we can just catch `ImportError` in addition to checking the raised `UserWarning`), but I'm worried that re-raising with an error telling users to check the pandas version may not always be the correct fix. If some other dependency of modin is installed with the wrong version as well as pandas, then an `ImportError` might have a different meaning.

Any thoughts on how to handle this?

**Comment by sfc-gh-azhan** [comment]
I started https://ci-dev-142.int.snowflakecomputing.com/job/SnowparkPandasStoredProcPrecommitTest/158/. Let's see how it goes. @sfc-gh-joshi

File: tox.ini
Comment (sfc-gh-azhan) [inline tox.ini:70]: move one line up.

File: .github/workflows/daily_modin_precommit.yml
Comment (sfc-gh-azhan) [inline .github/workflows/daily_modin_precommit.yml:149]: track with a jira no.

File: .github/workflows/daily_modin_precommit.yml
Comment (sfc-gh-azhan) [inline .github/workflows/daily_modin_precommit.yml:188]: why test `2.2.2` again here? Can we skip it?

Comment (sfc-gh-azhan) [inline .github/workflows/daily_modin_precommit.yml:188]: track with jira no.

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-azhan) [inline tests/integ/modin/test_modin_stored_procedures.py:33]: Did we do this because we want the current tests to pass? Is this supposedly the way our customer to use sproc?

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-azhan) [inline tests/integ/modin/test_modin_stored_procedures.py:284]: track with jira no.

**Review by sfc-gh-azhan — CHANGES_REQUESTED** [review]

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-joshi) [reply tests/integ/modin/test_modin_stored_procedures.py:33]: Customers should not need to do this, but this is necessary for tests to pass currently. As the comment says, even though `modin==0.30.1` is enabled for the testing account, since the latest version of `snowflake-snowpark-python` in anaconda pins `modin==0.28.1`, if we try to specify different modin version, the solver will resolve to an older version of the package that doesn't pin this dependency.

As long as the modin version is strictly pinned with `==` rather than `>=`, this will continue to be a problem moving forward whenever we try to upgrade the modin version.

**Comment by sfc-gh-azhan** [comment]
The precommit test failed https://ci-dev-142.int.snowflakecomputing.com/job/RT-Snowfort-Test-Custom/159/console

File: .github/workflows/daily_modin_precommit.yml
Comment (sfc-gh-yzou) [reply .github/workflows/daily_modin_precommit.yml:188]: For UDF and stored sproc it is expected, you can probably skip the udf and stored sproc test on pandas 2.2.3 for now, but for general tests, it should work

**Comment by sfc-gh-joshi** [comment]
> The precommit test failed https://ci-dev-142.int.snowflakecomputing.com/job/RT-Snowfort-Test-Custom/159/console

It looks like the failing test (`TestSnowparkPandasInSpPrecommit::test_modin_integ_precommit`) is still specifying `modin==0.28.1`. Should I open a PR to update the tests in Snowfort so it can be merged after this one?

**Comment by sfc-gh-azhan** [comment]
> > The precommit test failed https://ci-dev-142.int.snowflakecomputing.com/job/RT-Snowfort-Test-Custom/159/console
> 
> It looks like the failing test (`TestSnowparkPandasInSpPrecommit::test_modin_integ_precommit`) is still specifying `modin==0.28.1`. Should I open a PR to update the tests in Snowfort so it can be merged after this one?

All tests under https://github.com/snowflakedb/snowflake/tree/main/Snowfort/tests/t_snowpark_pandas/src are using 0.28.1 right now. How to update them without blocking precommit and this PR?

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/__init__.py:153]: and we have both pd.Dataframe.modin, and pd.Series.modin, right? otherwise, i don't think we will need to skip this for dir(Series)

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-yzou) [reply tests/integ/modin/test_modin_stored_procedures.py:45]: that seems something new you added in this test? why we start adding this change here now? does the test fails?

**Comment by sfc-gh-yzou** [comment]
@sfc-gh-azhan @sfc-gh-joshi what is the snowpark version used in TestSnowparkPandasInSpPrecommit, is it the most recent available version on snowflake anaconda or the master? if it is master, can we read the modin dependency and use that version for testing?

**Comment by sfc-gh-yzou** [comment]
and @sfc-gh-joshi have we tested this with our stored sproc jenkins job  https://ci-dev-142.int.snowflakecomputing.com/job/SnowparkPandasStoredProcPrecommitTest/

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-joshi) [reply src/snowflake/snowpark/modin/plugin/__init__.py:27]: Filed SNOW-1758773 to track

File: src/snowflake/snowpark/modin/plugin/extensions/base_overrides.py
Comment (sfc-gh-joshi) [reply src/snowflake/snowpark/modin/plugin/extensions/base_overrides.py:1561]: Filed SNOW-1758721

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-joshi) [reply tests/integ/modin/test_modin_stored_procedures.py]: I discussed with @sfc-gh-azhan, and I've reverted this change so we're just use the `snowflake-snowpark-python` + `pandas` + `modin` versions currently available in Anaconda.

File: tests/integ/modin/test_modin_stored_procedures.py
Comment (sfc-gh-joshi) [reply tests/integ/modin/test_modin_stored_procedures.py]: Discussed on Slack--I reverted this to just use latest `snowflake-snowpark-python`, and pin `pandas==2.2.1` and `modin==0.28.1` in this test (tracked by SNOW-1758760).

File: src/snowflake/snowpark/modin/plugin/__init__.py
Comment (sfc-gh-joshi) [reply src/snowflake/snowpark/modin/plugin/__init__.py:158]: Yes, `pd.DataFrame.modin` and `pd.Series.modin` are both valid; the `for attr_name in dir(ModinAPI)` loop will add telemetry to chained method accesses like `pd.DataFrame().modin.to_pandas()`, but we skip it in `dir(Series)` because our telemetry code has trouble handling the object returned by `pd.DataFrame().modin` (without a method call at the end).

File: .github/workflows/daily_modin_precommit.yml
Comment (sfc-gh-joshi) [reply .github/workflows/daily_modin_precommit.yml]: I removed 2.2.3 from this list.

pandas version updates happen infrequently enough that I think it would be fine to track this manually.

**Review by sfc-gh-azhan — APPROVED** [review]

**Review by sfc-gh-jjiao — APPROVED** [review]

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM. Thanks for keeping our dependencies up-to-date. (But please resolve all open conversations before merging.)

**Review by sfc-gh-yzou — APPROVED** [review]

**Comment by sfc-gh-joshi** [comment]
(from discussion on slack) Holding this PR until after 1.24 is cut.

**Pull request merged**
